pulsar-broker/src/main/java/org/apache/pulsar/broker/service/SystemTopicBasedLedgerDeletionService.java

```diff
@@ -0,0 +1,583 @@
+                CompletableFuture<?> future = new CompletableFuture<>();
+                future.completeExceptionally(
+                        new IllegalArgumentException("The ledger " + ledgerId + " didn't offload."));
+                return future;
+            }
+            pendingDeleteLedger =
+                    new PendingDeleteLedgerInfo(topicName, component, type, ledgerId, context, properties);
+        }
+        return sendMessage(pendingDeleteLedger);
+    }
+
+    @Override
+    public void close() throws Exception {
+        asyncClose().get();
+    }
+
+    @Override
+    public CompletableFuture<?> asyncClose() {
+        if (readerFuture != null && !readerFuture.isCompletedExceptionally()) {
+            return readerFuture.thenCompose(SystemTopicClient.Reader::closeAsync);
+        }
+        return CompletableFuture.completedFuture(null);
+    }
+
+    @Override
+    public boolean isTopicTwoPhaseDeletionEnabled() {
+        return true;
+    }
+
+    private CompletableFuture<?> handleMessage(LedgerDeletionSystemTopicClient.PendingDeleteLedgerReader reader,
+                                               Message<PendingDeleteLedgerInfo> message) {
+        CompletableFuture<?> future = new CompletableFuture<>();
+        deleteInBroker(message).whenComplete((res, ex) -> {
+            if (ex != null) {
+                if (ex instanceof PulsarAdminException.NotFoundException) {
+                    deleteLocally(message).whenComplete((res1, ex1) -> {
+                        if (ex1 != null) {
+                            if (ex1 instanceof PendingDeleteLedgerInvalidException) {
+                                if (log.isDebugEnabled()) {
+                                    log.debug("Received invalid pending delete ledger {}, invalid reason: {}",
+                                            message.getValue().getLedgerId(), ex1.getMessage());
+                                }
+                                reader.ackMessageAsync(message);
+                                future.complete(null);
+                                return;
+                            }
+                            reader.reconsumeLaterAsync(message);
+                            future.completeExceptionally(ex1);
+                        }
+                        reader.ackMessageAsync(message);
+                        future.complete(null);
+                    });
+                } else if (ex instanceof PendingDeleteLedgerInvalidException) {
+                    if (log.isDebugEnabled()) {
+                        log.debug("Received invalid pending delete ledger {}, invalid reason: {}",
+                                message.getValue().getLedgerId(), ex.getMessage());
+                    }
+                    reader.ackMessageAsync(message);
+                    future.complete(null);
+                } else if (ex instanceof PulsarAdminException.ServerSideErrorException) {
+
```

Comment:
Do we need to deal with this exception?

Reply:
It needn't, The origin logic is: in client side, to check the ServerSideErrorException if contains `PendingDeleteLedgerInvalidException` info, it contains it, means that the pending delete message is invalid, we also need ack it. 
But now, at server side, we catch `PendingDeleteLedgerInvalidException`, make it return success response. So the client side needn't to check the error response info